examples_tests/common/CommonAPI.h

```diff
@@ -443,128 +681,256 @@ class CommonAPI
+			result.apiConnection = _apiConnection;
+		}
+		else if(api_type == EAT_OPENGL)
+		{
+			auto _apiConnection = video::COpenGLConnection::create(core::smart_refctd_ptr(result.system), 0, app_name.data(), video::COpenGLDebugCallback(core::smart_refctd_ptr(result.logger)));
+			result.surface = video::CSurfaceGLWin32::create(core::smart_refctd_ptr(_apiConnection), core::smart_refctd_ptr<ui::IWindowWin32>(static_cast<ui::IWindowWin32*>(result.window.get())));
+			result.apiConnection = _apiConnection;
+		}
+		else if(api_type == EAT_OPENGL_ES)
+		{
+			auto _apiConnection = video::COpenGLESConnection::create(core::smart_refctd_ptr(result.system), 0, app_name.data(), video::COpenGLDebugCallback(core::smart_refctd_ptr(result.logger)));
+			result.surface = video::CSurfaceGLWin32::create(core::smart_refctd_ptr(_apiConnection), core::smart_refctd_ptr<ui::IWindowWin32>(static_cast<ui::IWindowWin32*>(result.window.get())));
+			result.apiConnection = _apiConnection;
+		}
+		else
+		{
+			_NBL_TODO();
+		}
 
-			if (graphicsQueueEnable)
-				requiredQueueFlags |= nbl::video::IPhysicalDevice::EQF_GRAPHICS_BIT;
+		auto gpus = result.apiConnection->getPhysicalDevices();
+		assert(!gpus.empty());
+		auto extractedInfos = extractGPUInfos(gpus, result.surface);
+		auto suitableGPUIndex = findSuitableGPU(extractedInfos, graphicsQueueEnable);
+		auto gpu = gpus.begin()[suitableGPUIndex];
+		const auto& gpuInfo = extractedInfos[suitableGPUIndex];
+
+		float queuePriority = 1.f;
+		constexpr uint32_t MaxQueueCount = 4;
+		video::ILogicalDevice::SQueueCreationParams qcp[MaxQueueCount] = {}; 
+		
+		uint32_t actualQueueCount = 1;
+		uint32_t mainQueueFamilyIndex = QueueFamilyProps::InvalidIndex;
+		if(graphicsQueueEnable)
+			mainQueueFamilyIndex = gpuInfo.queueFamilyProps.graphics.index;
+		else
+			mainQueueFamilyIndex = gpuInfo.queueFamilyProps.compute.index;
 
-			return getQueueFamilyIndex(gpu, requiredQueueFlags);
-		};
+		qcp[0].familyIndex = mainQueueFamilyIndex;
+		qcp[0].count = 1u;
+		qcp[0].flags = static_cast<video::IGPUQueue::E_CREATE_FLAGS>(0);
+		qcp[0].priorities = &queuePriority;
 
-		const auto familyIndex = getFamilyIndex();
+		if(qcp[0].familyIndex != gpuInfo.queueFamilyProps.compute.index)
+		{
+			qcp[actualQueueCount].flags = static_cast<video::IGPUQueue::E_CREATE_FLAGS>(0);
+			qcp[actualQueueCount].familyIndex = gpuInfo.queueFamilyProps.compute.index;
+			qcp[actualQueueCount].count = 1u;
+			qcp[actualQueueCount].priorities = &queuePriority;
+			actualQueueCount++;
+		}
+		if(gpuInfo.queueFamilyProps.transfer.index != gpuInfo.queueFamilyProps.compute.index && gpuInfo.queueFamilyProps.transfer.index != gpuInfo.queueFamilyProps.graphics.index)
+		{
+			qcp[actualQueueCount].flags = static_cast<video::IGPUQueue::E_CREATE_FLAGS>(0);
+			qcp[actualQueueCount].familyIndex = gpuInfo.queueFamilyProps.transfer.index;
+			qcp[actualQueueCount].count = 1u;
+			qcp[actualQueueCount].priorities = &queuePriority;
+			actualQueueCount++;
+		}
```

Comment:
I think I'll always want 2 separate queues for upload and download even if they're from the same family

Reply:
I will write code that tries to achieve this. But note that there can be a case where there is only 1 queue family and maxQueueCount = 1 which supports all (I even have the device, It's intel UHD something)
I'm on it.